sale_commission_product_criteria_domain/models/partner.py

```diff
@@ -40,14 +40,19 @@ def _onchange_agent_ids(self):
         for rec in self:
             exiting_agents = rec.commission_item_agent_ids.mapped("agent_id")
             to_create = [
-                {"partner_id": rec._origin.id, "agent_id": x._origin.id}
+                {
+                    "partner_id": rec._origin.id,
+                    "agent_id": x._origin.id,
+                    "group_ids": [(6, 0, x.allowed_commission_group_ids.ids)],
+                }
                 for x in rec.agent_ids.filtered(
                     lambda x: x.commission_id.commission_type == "product_restricted"
                 )
-                if x not in exiting_agents.ids
+                if x._origin.id not in exiting_agents.ids
             ]
             to_delete = rec.commission_item_agent_ids.filtered(
-                lambda x: x.agent_id.id in (exiting_agents - rec.agent_ids).ids
+                lambda x: x.agent_id._origin.id
+                in (set(exiting_agents.ids) - set(rec.agent_ids.ids))
```

Comment:
Maybe there's something I'm missing, but why not work with recordsets directly? 

```suggestion
                lambda x: x.agent_id in (exiting_agents - rec.agent_ids)
```

Reply:
Same here. It cant be done using records. Thats why i use ids and sets.

Reply:
This seems to work for me, on both existing partners and partners being created. Could you double check please?

```py
    @api.onchange("agent_ids")
    def _onchange_agent_ids(self):
        for rec in self:
            exiting_agents = rec.commission_item_agent_ids.mapped("agent_id")
            to_create = [
                {
                    "agent_id": x.id,
                    "group_ids": [(6, 0, x.allowed_commission_group_ids.ids)],
                }
                for x in rec.agent_ids.filtered(
                    lambda x: x.commission_id.commission_type == "product_restricted"
                )
                if x not in exiting_agents
            ]
            to_delete = rec.commission_item_agent_ids.filtered(
                lambda x: x.agent_id in (exiting_agents - rec.agent_ids)
            )
            if to_delete:
                rec.update(
                    {"commission_item_agent_ids": [(2, dl.id, 0) for dl in to_delete]}
                )
            if to_create:
                rec.update(
                    {"commission_item_agent_ids": [(0, 0, line) for line in to_create]}
                )
```

If you confirm that it doesn't work for you I'll drop the issue, since it's no big deal either way, but I'm curious to understand why it wouldn't work.

Reply:
It works indeed. Must be my mistake. Thanks!